File: backend/src/main/java/ar/com/rotula/domain/Product.java
package ar.com.rotula.domain;

import jakarta.persistence.*;
import lombok.*;

import java.math.BigDecimal;
import java.time.OffsetDateTime;
import java.util.UUID;

@Entity
@Table(name = "products")
@Getter @Setter @NoArgsConstructor @AllArgsConstructor @Builder
public class Product {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private UUID id;

    @Column(name = "tenant_id", nullable = false)
    private UUID tenantId;

    /** Nombre interno del producto (no aparece en el rótulo). */
    @Column(nullable = false, length = 300)
    private String name;

    /**
     * Denominación del alimento según el CAA (aparece en el rótulo).
     * Ej.: "Galletitas dulces de avena".
     */
    @Column(length = 300)
    private String denomination;

    // ── Clasificación por TABLA I ─────────────────────────────────────────

    /**
     * Grupo de alimentos de la TABLA I (Res. Conjunta 21/2023).
     * Reemplaza al campo category — NO aparece en el rótulo.
     */
    @Column(name = "food_group_id")
    private UUID foodGroupId;

    /**
     * Alimento específico dentro del grupo.
     * Determina la porción de referencia (serving_size_g).
     */
    @Column(name = "food_item_id")
    private UUID foodItemId;

    // ── Presentación ──────────────────────────────────────────────────────

    @Column(name = "net_weight", nullable = false, precision = 10, scale = 3)
    private BigDecimal netWeight;

    @Column(name = "weight_unit", nullable = false, length = 10)
    private String weightUnit;

    /**
     * Tamaño de porción en gramos (o mL para líquidos).
     * Se auto-completa desde food_item.portion_grams al seleccionar el alimento.
     * No es editable por el usuario.
     */
    @Column(name = "serving_size_g", precision = 8, scale = 2)
    private BigDecimal servingSizeG;

    // ── Registros ─────────────────────────────────────────────────────────

    @Column(name = "rne_number", length = 20)
    private String rneNumber;

    /** Obligatorio desde V8. Número de Registro Nacional de Producto Alimenticio. */
    @Column(name = "rnpa_number", length = 20)
    private String rnpaNumber;

    // ── Alérgenos / contaminación cruzada ─────────────────────────────────

    /**
     * Grupos de alérgenos presentes en el ambiente de producción.
     * Se almacena como nombres del enum {@link AllergenGroup} separados por coma.
     * Ejemplo: "MANI,FRUTOS_DE_CASCARA"
     */
    @Column(name = "cross_contamination")
    private String crossContamination;

    // ── Opciones de rótulo ────────────────────────────────────────────────

    /**
     * Si true, el % de cada ingrediente se muestra en la lista de ingredientes
     * del rótulo (es optativo según la normativa vigente).
     */
    @Column(name = "show_ingredient_percentages", nullable = false)
    private boolean showIngredientPercentages;

    // ── Tabla nutricional por 100 g ───────────────────────────────────────
    // El sistema calcula los valores por porción = campo * serving_size_g / 100

    @Column(name = "energy_kcal_per100g", precision = 8, scale = 2)
    private BigDecimal energyKcalPer100g;

    @Column(name = "proteins_g_per100g", precision = 8, scale = 2)
    private BigDecimal proteinsPer100g;

    @Column(name = "carbs_g_per100g", precision = 8, scale = 2)
    private BigDecimal carbsPer100g;

    @Column(name = "sugars_g_per100g", precision = 8, scale = 2)
    private BigDecimal sugarsPer100g;

    @Column(name = "fat_total_g_per100g", precision = 8, scale = 2)
    private BigDecimal fatTotalPer100g;

    @Column(name = "fat_sat_g_per100g", precision = 8, scale = 2)
    private BigDecimal fatSatPer100g;

    @Column(name = "fat_trans_g_per100g", precision = 8, scale = 2)
    private BigDecimal fatTransPer100g;

    @Column(name = "sodium_mg_per100g", precision = 8, scale = 2)
    private BigDecimal sodiumMgPer100g;

    // ── Auditoría ─────────────────────────────────────────────────────────

    @Column(length = 20)
    private String status;

    @Column(name = "created_by")
    private UUID createdBy;

    @Column(name = "created_at", nullable = false)
    private OffsetDateTime createdAt;

    @Column(name = "updated_at", nullable = false)
    private OffsetDateTime updatedAt;

    @PrePersist
    void onCreate() {
        createdAt = updatedAt = OffsetDateTime.now();
    }

    @PreUpdate
    void onUpdate() {
        updatedAt = OffsetDateTime.now();
    }
}

File: frontend/src/types/product.ts
// ── Datos de referencia TABLA I ───────────────────────────────────────────────

export interface FoodGroup {
  id:        string
  code:      string
  name:      string
  sortOrder: number
}

export interface FoodItem {
  id:           string
  foodGroupId:  string
  name:         string
  /** Porción de referencia en gramos o mL (TABLA I – Res. Conjunta 21/2023). */
  portionGrams: number
  /** Unidad de la porción: 'g' o 'ml'. */
  unit:         string
  sortOrder:    number
}

// ── Producto ──────────────────────────────────────────────────────────────────

export interface Product {
  id:        string
  tenantId:  string

  /** Nombre interno (no aparece en el rótulo). */
  name:        string
  /** Denominación del alimento según CAA (aparece en el rótulo). */
  denomination: string

  // Clasificación TABLA I
  foodGroupId: string
  foodItemId:  string
  /** Porción de referencia (auto-completada desde el food_item seleccionado). */
  servingSizeG: number

  // Presentación
  netWeight:   number
  weightUnit:  string

  // Registros
  rneNumber:  string | null
  rnpaNumber: string       // obligatorio

  // Contaminación cruzada (lista de nombres del enum AllergenGroup)
  crossContaminationGroups: string[]

  // Opciones de rótulo
  showIngredientPercentages: boolean

  // Tabla nutricional por 100 g (todos opcionales)
  energyKcalPer100g: number | null
  proteinsPer100g:   number | null
  carbsPer100g:      number | null
  sugarsPer100g:     number | null
  fatTotalPer100g:   number | null
  fatSatPer100g:     number | null
  fatTransPer100g:   number | null
  sodiumMgPer100g:   number | null

  status:    string
  createdBy: string
  createdAt: string
  updatedAt: string
}

export interface ProductRequest {
  name:         string
  denomination: string

  foodGroupId: string
  foodItemId:  string
  servingSizeG?: number | null

  netWeight:  number
  weightUnit: string

  rneNumber?:  string | null
  rnpaNumber:  string

  crossContaminationGroups: string[]
  showIngredientPercentages: boolean

  energyKcalPer100g?: number | null
  proteinsPer100g?:   number | null
  carbsPer100g?:      number | null
  sugarsPer100g?:     number | null
  fatTotalPer100g?:   number | null
  fatSatPer100g?:     number | null
  fatTransPer100g?:   number | null
  sodiumMgPer100g?:   number | null
}

export interface PageResponse<T> {
  content:       T[]
  page:          number
  size:          number
  totalElements: number
  totalPages:    number
  last:          boolean
}

// ── Label analysis types ──────────────────────────────────────────────────────

export interface RoundedNutritionValues {
  energyKcal: number
  energyKj:   number
  proteinsG:  number
  carbsG:     number
  sugarsG:    number
  fatTotalG:  number
  fatSatG:    number
  fatTransG:  number
  sodiumMg:   number
}

export interface NutritionValues extends RoundedNutritionValues {}

export interface NutritionCalculationResult {
  servingSizeG: number
  perPortion:   RoundedNutritionValues
  per100g:      RoundedNutritionValues
  rawPer100g:   NutritionValues
}

export interface AllergenDeclarationResult {
  presentGroups:            string[]
  crossContaminationGroups: string[]
  declarationText:          string
}

export interface LabelAnalysisResult {
  productId: string
  nutrition: NutritionCalculationResult | null
  seals:     string[]
  allergens: AllergenDeclarationResult
}
